src/core/reactor_backend.cc

```diff
@@ -1478,51 +1493,150 @@ class reactor_backend_uring final : public reactor_backend {
         fd.fd.shutdown(how);
     }
     virtual future<size_t> read_some(pollable_fd_state& fd, void* buffer, size_t len) override {
-        return readable(fd).then([this, &fd, buffer, len] () mutable {
-            auto r = fd.fd.read(buffer, len);
-            if (!r) {
-                return read_some(fd, buffer, len);
-            }
-            if (size_t(*r) == len) {
-                fd.speculate_epoll(EPOLLIN);
+        if (fd.take_speculation(POLLIN)) {
+            try {
+                auto r = fd.fd.read(buffer, len);
+                if (r) {
+                    if (size_t(*r) == len) {
+                        fd.speculate_epoll(EPOLLIN);
+                    }
+                    return make_ready_future<size_t>(*r);
+                }
+            } catch (...) {
+                return current_exception_as_future<size_t>();
             }
-            return make_ready_future<size_t>(*r);
+        }
+        return readable(fd).then([this, &fd, buffer, len] () mutable {
+            class read_completion final : public io_completion {
+                pollable_fd_state& _fd;
+                const size_t _to_read;
+                promise<size_t> _result;
+            public:
+                read_completion(pollable_fd_state& fd, size_t to_read)
+                    : _fd(fd), _to_read(to_read) {}
+                void complete(size_t bytes) noexcept final {
+                    if (bytes == _to_read) {
+                        _fd.speculate_epoll(EPOLLIN);
+                    }
+                    _result.set_value(bytes);
+                    delete this;
+                }
+                void set_exception(std::exception_ptr eptr) noexcept final {
+                    _result.set_exception(eptr);
+                    delete this;
+                }
+                future<size_t> get_future() {
+                    return _result.get_future();
+                }
+            };
+            auto desc = std::make_unique<read_completion>(fd, len);
+            auto req = internal::io_request::make_read(fd.fd.get(), -1, buffer, len, false);
+            return submit_request(std::move(desc), std::move(req));
```

Comment:
I think we need to keep this as the original loop, since the new code takes two round-trips in the ring, so has higher latency.

Although, since it's just for inotify, it doesn't matter much.

Reply:
as io_uring `read` op will retry automatically, i think we can dispense with the `readable(fd).then` part. so in the worst case, just one round trip would suffice.

anyway, will add this change in a follow-up PR.